vault/raft.go

```diff
@@ -524,18 +534,29 @@ func (c *Core) raftSnapshotRestoreCallback(grabLock bool, sealNode bool) func(co
 	}
 }
 
-func (c *Core) JoinRaftCluster(ctx context.Context, leaderAddr string, tlsConfig *tls.Config, retry, nonVoter bool) (bool, error) {
-	if len(leaderAddr) == 0 {
-		return false, errors.New("No leader address provided")
+func (c *Core) InitiateRetryJoin(ctx context.Context) (bool, error) {
+	raftStorage, ok := c.underlyingPhysical.(*raft.RaftBackend)
+	if !ok {
+		return false, errors.New("raft storage not configured")
+	}
+
+	retryJoinConfig := raftStorage.RetryJoinConfig()
+	if retryJoinConfig == "" {
+		return false, nil
 	}
 
+	c.logger.Info("raft retry join initiated")
+	return c.JoinRaftCluster(ctx, nil, false)
```

Comment:
We are actually hinging on the fact that leaderInfo is nil, to know whether or not to pull the config from the backend. This should be nil as we will not know leader info at this point.

Reply:
Looks like you are now calling `raftStorage.JoinConfig(nil)` multiple times. Why can't we just call it once outside of the function?

Reply:
We are calling it only once. Am I missing something?

Reply:
You're calling it once here: 

https://github.com/hashicorp/vault/blob/2f8b484252d11e8955aeb2a190c7bb7282367780/vault/raft.go#L532-L535

which then goes on to call `JoinRaftCluster`

Then its getting called a second time in `JoinRaftCluster` here: 

https://github.com/hashicorp/vault/blob/2f8b484252d11e8955aeb2a190c7bb7282367780/vault/raft.go#L567-L570

I think this should only be called once outside of `JoinRaftCluster` and the result should be passed in as a parameter. This would likely mean changing the signature from:
```
JoinRaftCluster(ctx context.Context, leaderInfo *raft.LeaderJoinInfo, nonVoter bool) (bool, error) {
```
To:
```
JoinRaftCluster(ctx context.Context, leaderInfos []*raft.LeaderJoinInfo, nonVoter bool) (bool, error) {
```

Reply:
Ah, I see. I misread you meant that we are calling it in a loop. Sure, that can be done.

Reply:
Done